File: go/ch01/greedy_intro_hc.go
// go/ch01/greedy_intro_hc.go
package main

// 各日 26 通り試して、その日終了時満足度が最大のタイプを選ぶ（骨格）
func greedySchedule(d int, c []int, s [][]int, scoreFn func([]int) []int) []int {
	schedule := make([]int, 0, d)
	for day := 1; day <= d; day++ {
		bestT, bestSat := 1, int(-1e18)
		for t := 1; t <= 26; t++ {
			trial := append(append([]int{}, schedule...), t)
			sat := scoreFn(trial)[len(trial)-1]
			if sat > bestSat {
				bestSat = sat
				bestT = t
			}
		}
		schedule = append(schedule, bestT)
	}
	return schedule
}

File: python/ch01/greedy_intro_hc.py
# python/ch01/greedy_intro_hc.py
"""Intro HC 問題A 向け貪欲の骨格（第5章で本実装）."""


def daily_satisfactions(d, c, s, schedule):
    last = [0] * 26
    satisfaction = 0
    result = []
    for day in range(1, d + 1):
        t = schedule[day - 1] - 1
        satisfaction += s[day - 1][t]
        last[t] = day
        decay = sum(c[i] * (day - last[i]) for i in range(26))
        satisfaction -= decay
        result.append(satisfaction)
    return result


def greedy_schedule(d, c, s):
    schedule = []
    for day in range(1, d + 1):
        best_t, best_sat = 1, None
        for t in range(1, 27):
            trial = schedule + [t]
            sat = daily_satisfactions(day, c, s, trial)[-1]
            if best_sat is None or sat > best_sat:
                best_sat = sat
                best_t = t
        schedule.append(best_t)
    return schedule
